desc/integrals/quad_utils.py

```diff
@@ -0,0 +1,247 @@
+
+
+grad_automorphism_arcsin.__doc__ += "\n" + automorphism_arcsin.__doc__
+
+
+def automorphism_sin(x, s=0, m=10):
+    """[-1, 1] ∋ x ↦ y ∈ [−1, 1].
+
+    When used as the change of variable map for the bounce integral, the Lipschitzness
+    of the sin transformation prevents generation of new singularities. Furthermore,
+    its derivative vanishes to zero slowly near the boundary, which will suppress the
+    large derivatives near the boundary of singular integrals.
+
+    In effect, this map pulls the mass of the integral away from the singularities,
+    which should improve convergence if the quadrature performs better on less singular
+    integrands. Pairs well with Gauss-Legendre quadrature.
+
+    Parameters
+    ----------
+    x : jnp.ndarray
+        Points to transform.
+    s : float
+        Strength of derivative suppression, s ∈ [0, 1].
+    m : float
+        Number of machine epsilons used for floating point error buffer.
+
+    Returns
+    -------
+    y : jnp.ndarray
+        Transformed points.
+
+    """
+    errorif(not (0 <= s <= 1))
+    # s = 0 -> derivative vanishes like cosine.
+    # s = 1 -> derivative vanishes like cosine^k.
+    y0 = jnp.sin(0.5 * jnp.pi * x)
+    y1 = x + jnp.sin(jnp.pi * x) / jnp.pi  # k = 2
+    y = (1 - s) * y0 + s * y1
+    # y is an expansion, so y(x) > x near x ∈ {−1, 1} and there is a tendency
+    # for floating point error to overshoot the true value.
+    eps = m * jnp.finfo(jnp.array(1.0).dtype).eps
+    return jnp.clip(y, -1 + eps, 1 - eps)
+
+
+def grad_automorphism_sin(x, s=0):
+    """Gradient of sin automorphism."""
+    dy0_dx = 0.5 * jnp.pi * jnp.cos(0.5 * jnp.pi * x)
+    dy1_dx = 1.0 + jnp.cos(jnp.pi * x)
+    dy_dx = (1 - s) * dy0_dx + s * dy1_dx
+    return dy_dx
+
+
+grad_automorphism_sin.__doc__ += "\n" + automorphism_sin.__doc__
+
+
+def tanh_sinh(deg, m=10):
+    """Tanh-Sinh quadrature.
+
+    Returns quadrature points xₖ and weights wₖ for the approximate evaluation of the
+    integral ∫₋₁¹ f(x) dx ≈ ∑ₖ wₖ f(xₖ).
+
+    Parameters
+    ----------
+    deg : int
+        Number of quadrature points.
+    m : float
+        Number of machine epsilons used for floating point error buffer. Larger implies
+        less floating point error, but increases the minimum achievable error.
+
+    Returns
+    -------
+    x, w : (jnp.ndarray, jnp.ndarray)
+        Shape (deg, ).
+        Quadrature points and weights.
+
+    """
+    # buffer to avoid numerical instability
+    x_max = jnp.array(1.0)
+    x_max = x_max - m * jnp.finfo(x_max.dtype).eps
+    t_max = jnp.arcsinh(2 * jnp.arctanh(x_max) / jnp.pi)
+    # maximal-spacing scheme, doi.org/10.48550/arXiv.2007.15057
+    t = jnp.linspace(-t_max, t_max, deg)
+    dt = 2 * t_max / (deg - 1)
+    arg = 0.5 * jnp.pi * jnp.sinh(t)
+    x = jnp.tanh(arg)  # x = g(t)
+    w = 0.5 * jnp.pi * jnp.cosh(t) / jnp.cosh(arg) ** 2 * dt  # w = (dg/dt) dt
+    return x, w
+
+
+def leggauss_lob(deg, interior_only=False):
```

Comment:
The integrals with factors $\sqrt(1 - \lambda \vert B\vert)$, like all the ones in the effective ripple, vanish at the endpoints. When we use the $N$ interior points of a lobatto quadrature of degree $N+2$, we can integrate polynomials of degree $2(N + 2) - 3 = 2N + 1$ exactly. An $N$ point non-lobatto quadrature caps out at $2N - 1$. 

1. After the change of variables to remove the singularity, the integrands are polynomial-like, so this +2 boost is not negligible.
2. There are more careful ways to comparing Gauss-Lobatto and Gauss quadratures. Recall, these quadratures interpolating a polynomial to the quadrature points and integrating that exactly. A lobatto quadrature will fit at the endpoints and help avoid runge, so we should get a better improvement than just the +2 indicates.
3. The interior node closest to the boundary in a Gauss-Lobatto quadrature is always further from the boundary than it is in Gauss quadrature. This will make the integration more resistant to floating point errors in computing bounce points.